### PR TITLE
brig: Add exact handle matches from all teams in /search/contacts

### PR DESCRIPTION
**Pull request by akshaymankar**
Fixes https://wearezeta.atlassian.net/browse/SQCORE-563

Pending:

* [x] Enable doing federated exact handle search
* [x] Mark `/users/by-handle/:domain/:handle` endpoint as deprecated (can we delete it already?)

### CODE REVIEW

File: services/brig/src/Brig/User/API/Search.hs [hunk 1]
```diff
@@ -132,30 +134,36 @@ routesInternal = do
 
 -- Handlers
 
+-- TODO question: would it make sense to create a second endpoint with a mandatory domain?
+-- That would make it easier to see from logs whether the old endpoint is still in use or can be removed.
+--
+-- TODO question: would it make sense to augment the 'SearchResult'
+-- to include a separate entry for exact handle match? This may allow clients to handle that differently.
+-- Not sure if useful or not - perhaps discuss with clients?
```
Comment (akshaymankar) [inline services/brig/src/Brig/User/API/Search.hs:142]: Seems like there is already demand for this: https://wearezeta.atlassian.net/browse/BE-262

File: libs/wire-api/src/Wire/API/User/Search.hs [hunk 2]
```diff
@@ -132,7 +182,7 @@ instance FromJSON Contact where
   parseJSON =
     withObject "Contact" $ \o ->
       Contact
-        <$> o .: "id"
+        <$> o .: "qualified_id"
```
[the remark this answers is not in the document]
Comment (akshaymankar) [reply libs/wire-api/src/Wire/API/User/Search.hs:185]: We never read this from outside, this `FromJSON` only exists for our tests, so I think it is fine to break backwards compatibility here.

[the remark this answers is not in the document]
Comment (jschaul) [reply libs/wire-api/src/Wire/API/User/Search.hs:185]: Okay, sure!

File: libs/wire-api/src/Wire/API/User/Search.hs [hunk 3]
```diff
@@ -102,6 +136,21 @@ data Contact = Contact
   deriving stock (Eq, Show, Generic)
   deriving (Arbitrary) via (GenericUniform Contact)
 
+-- deriving (ToSchema) via (CustomSwagger '[FieldLabelModifier (StripPrefix "contact", CamelToSnake, LabelMappings ContactLabelMappings)] Contact)
+
```
Comment (akshaymankar) [inline libs/wire-api/src/Wire/API/User/Search.hs:140]:
```suggestion
```

File: services/brig/src/Brig/User/Search/SearchIndex.hs [hunk 4]
```diff
@@ -107,8 +121,12 @@ defaultUserQuery u teamSearchInfo (normalized -> term') =
           boolQuery
             { ES.boolQueryMustMatch =
                 [ ES.QueryBoolQuery
-                    boolQuery {ES.boolQueryShouldMatch = [matchPhraseOrPrefix, legacyPrefixMatch]}
+                    boolQuery
+                      { ES.boolQueryShouldMatch = [matchPhraseOrPrefix, legacyPrefixMatch],
+                        ES.boolQueryMustNotMatch = [termQ "handle" term']
+                      }
                 ],
+              -- This removes exact handle matches, as they are fetched from cassandra
```
Comment (akshaymankar) [inline services/brig/src/Brig/User/Search/SearchIndex.hs:129]:
```suggestion
                    boolQuery
                      { ES.boolQueryShouldMatch = [matchPhraseOrPrefix, legacyPrefixMatch],
                        -- This removes exact handle matches, as they are fetched from cassandra
                        ES.boolQueryMustNotMatch = [termQ "handle" term']
                      }
                ],
```

The comment is at wrong place, either I was being absent minded or ormolu did something weird!

File: services/brig/test/integration/API/Federation.hs [hunk 5]
```diff
@@ -17,37 +17,69 @@
...
 import Util
 
+-- FUTUREWORK(federation): use servant-client in tests for the federation endpoints instead of the bilge requests.
 tests :: Manager -> Brig -> IO TestTree
 tests m brig = do
   return $
     testGroup "federation" $
-      [ test m "GET /federation/users/by-handle : 200" (testGetUserByHandleSuccess brig),
+      [ test m "GET /federation/search/users : 200" (testSearchSuccess brig),
+        test m "GET /federation/search/users : 404" (testSearchNotFound brig),
```
Comment (akshaymankar) [inline services/brig/test/integration/API/Federation.hs:40]:
```suggestion
      [ test m "GET /federation/search/users : Found" (testSearchSuccess brig),
        test m "GET /federation/search/users : NotFound" (testSearchNotFound brig),
```

As the HTTP status for both is 200.

**Review by akshaymankar — COMMENTED** [review]
Looks good! I have added some minor suggestions. I cannot approve this as I started the PR, so feel free to merge!

**Comment by akshaymankar** [comment]
Also, there are conflicts!

**Review by jschaul — APPROVED** [review]

**Pull request merged**